### PR TITLE
[jit] Fix reflection on weak modules, copy attributes

### LINKED ISSUE #19363: [JIT] Cannot access nn.Linear.in_features in ScriptModule
## 🐛 Bug

```python
import torch.nn as nn
import torch.jit as jit

class TestModule(jit.ScriptModule):
  def __init__(self):
    super().__init__()
    self.linear = nn.Linear(16, 16)


m = TestModule()
print(m.linear.in_features)
```

The code above throws AttributeError

```python
>>> m.linear.in_features
Traceback (most recent call last):
  File "/home/qbx2/anaconda3/lib/python3.7/site-packages/torch/jit/__init__.py", line 1197, in __getattr__
    return ScriptModule.__getattr__(self, attr)
  File "/home/qbx2/anaconda3/lib/python3.7/site-packages/torch/jit/__init__.py", line 1102, in __getattr__
    return Module.__getattr__(self, attr)
  File "/home/qbx2/anaconda3/lib/python3.7/site-packages/torch/nn/modules/module.py", line 535, in __getattr__
    type(self).__name__, name))
AttributeError: 'WeakScriptModuleProxy' object has no attribute 'in_features'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/qbx2/anaconda3/lib/python3.7/site-packages/torch/jit/__init__.py", line 1200, in __getattr__
    return getattr(self.__dict__["_original"](), attr)
AttributeError: 'NoneType' object has no attribute 'in_features'
```

## Expected behavior

Looking at the code (https://github.com/pytorch/pytorch/blob/master/torch/jit/__init__.py#L1226), I think WeakScriptModuleProxy should have copied in_features and other fields to self here.

## Environment

Collecting environment information...
PyTorch version: 1.0.1.post2
Is debug build: No
CUDA used to build PyTorch: 10.0.130

OS: Ubuntu 18.04.1 LTS
GCC version: (Ubuntu 7.3.0-27ubuntu1~18.04) 7.3.0
CMake version: Could not collect

Python version: 3.7
Is CUDA available: Yes
CUDA runtime version: 10.0.130
GPU models and configuration: GPU 0: GeForce RTX 2080
Nvidia driver version: 410.48
cuDNN version: /usr/local/cuda-10.0/lib64/libcudnn.so.7.4.1

Versions of relevant libraries:
[pip] numpy==1.15.4
[pip] numpydoc==0.8.0
[pip] torch==1.0.1.post2
[conda] blas 1.0 mkl
[conda] mkl 2019.1 144
[conda] mkl-service 1.1.2 py37he904b0f_5
[conda] mkl_fft 1.0.6 py37hd81dba3_0
[conda] mkl_random 1.0.2 py37hd81dba3_0
[conda] pytorch 1.0.1 py3.7_cuda10.0.130_cudnn7.4.2_2 pytorch

https://discuss.pytorch.org/t/cannot-access-nn-linear-in-features-in-scriptmodule/42833

### PR DESCRIPTION
**Pull request by driazati**
* Constructs a new type at runtime so that `isinstance` checks work for
weak modules assigned to `ScriptModule`s
* Fix some extraneous names in `__constants__`
* Add `in_features` and `out_features` to `nn.Linear` `__constants__`

Fixes #19363

Differential Revision: [D15302350](https://our.internmc.facebook.com/intern/diff/15302350/)

### CODE REVIEW

File: torch/jit/__init__.py [hunk 1]
```diff
@@ -1469,6 +1469,17 @@ def __init__(self, original, stubs):
             # Copy constants
             self.__dict__["_constants_set"] = set(getattr(original, "__constants__", []))
 
+            # Automatically register primitive types as attributes
+            for name in dir(original):
```
Comment (eellison) [inline torch/jit/__init__.py]: It seems strange that we are taking all values in dir(original) - won't be we be getting arbitrary python implementation details here ?

File: torch/jit/__init__.py [hunk 2]
```diff
@@ -1481,7 +1492,8 @@ def __getattr__(self, attr):
             try:
                 return ScriptModule.__getattr__(self, attr)
             except AttributeError:
-                if self.__dict__["_initialized"]:
+                orig = self.__dict__["_original"]()
```
Comment (eellison) [inline torch/jit/__init__.py]: Could you add a comment here? i don't understand what's going on

File: torch/nn/modules/linear.py [hunk 3]
```diff
@@ -136,7 +134,7 @@ class Bilinear(Module):
         >>> print(output.size())
         torch.Size([128, 40])
     """
-    __constants__ = ['in1_features', 'in2_features', 'out_features', 'bias']
+    __constants__ = ['in1_features', 'in2_features', 'out_features']
```
Comment (eellison) [inline torch/nn/modules/linear.py]: I don't think these changes are being tested

File: torch/jit/__init__.py [hunk 4]
```diff
@@ -1446,11 +1446,11 @@ class WeakScriptModuleProxy(ScriptModule):
         def __init__(self, original, stubs):
             # Guards behavior of __setattr__ and __getattr__ so ScriptModule
             # __init__ can run correctly
+            object.__setattr__(self, '__dict__', {})
```
Comment (eellison) [inline torch/jit/__init__.py]: What does this do?

File: torch/jit/__init__.py [hunk 5]
```diff
@@ -1460,25 +1464,19 @@ def __init__(self, original, stubs):
...
-            for name in dir(original):
-                if name == 'training':
-                    # TODO: remove this hack when training is a bool attribute
-                    # instead of a buffer
-                    continue
-                item = getattr(original, name)
-                primitive_type = torch.jit.annotations.get_primitive_type(item)
-                if name not in self.__dict__["_constants_set"] and primitive_type:
-                    ScriptModule.__setattr__(self, name, torch.jit.Attribute(item, primitive_type))
+            self.__dict__["_constants_set"] = constants_set
```
Comment (eellison) [inline torch/jit/__init__.py]: Could you move the copy constants logic below the `self.__dict__["_constants_set"] = {}` line ? I think that will read more clearly.

File: torch/nn/modules/rnn.py [hunk 6]
```diff
@@ -26,7 +26,7 @@ def apply_permutation(tensor, permutation, dim=1):
 
 class RNNBase(Module):
     __constants__ = ['mode', 'input_size', 'hidden_size', 'num_layers', 'bias',
-                     'batch_first', 'dropout', 'bidirectional', '_flat_parameters']
+                     'batch_first', 'dropout', 'bidirectional']
```
Comment (eellison) [inline torch/nn/modules/rnn.py]: what were reasons for these changes?

File: torch/nn/modules/linear.py [hunk 7]
```diff
@@ -67,6 +67,8 @@ class Linear(Module):
         >>> print(output.size())
         torch.Size([128, 30])
     """
+    __constants__ = ['bias', 'in_features', 'out_features']
```
Comment (eellison) [inline torch/nn/modules/linear.py]: Not related to this diff but: how did this work without constants previously? Should there have been a test failure somewhere that didn't happen?

File: torch/jit/__init__.py [hunk 8]
```diff
@@ -1492,9 +1490,11 @@ def __getattr__(self, attr):
             try:
                 return ScriptModule.__getattr__(self, attr)
             except AttributeError:
-                orig = self.__dict__["_original"]()
-                if orig and self.__dict__["_initialized"]:
-                    return getattr(self.__dict__["_original"](), attr)
+                # unwrap the original
+                original_module = self.__dict__["_original"]()
+                if original_module and self.__dict__["_initialized"]:
```
Comment (eellison) [inline torch/jit/__init__.py]: Why do we need the "initialized" check here?

**Review by eellison — COMMENTED** [review]
Looks good, have a couple of qauestions

File: torch/nn/modules/linear.py (hunk 7 above)
Comment (driazati) [reply torch/nn/modules/linear.py]: They're not actually used in `nn.Linear` except in `__repr__`

File: torch/nn/modules/rnn.py (hunk 6 above)
Comment (driazati) [reply torch/nn/modules/rnn.py]: This was unused

File: torch/jit/__init__.py (hunk 5 above)
Comment (driazati) [reply torch/jit/__init__.py]: This line has to happen after the parameters due to #14715

File: torch/jit/__init__.py (hunk 8 above)
Comment (driazati) [reply torch/jit/__init__.py]: So the `__getattr__` only looks at the original model after `__init__` is done

**Review by eellison — APPROVED** [review]

**Pull request closed without merging**